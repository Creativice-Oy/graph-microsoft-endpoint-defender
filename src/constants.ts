import {
  RelationshipClass,
  RelationshipDirection,
  StepEntityMetadata,
  StepMappedRelationshipMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export const Steps: Record<
  | 'FETCH_ACCOUNT'
  | 'FETCH_MACHINES'
  | 'FETCH_LOGON_USERS'
  | 'FETCH_USERS'
  | 'FETCH_FINDINGS'
  | 'FINDING_CVE_RELATIONSHIP'
  | 'FETCH_ENDPOINTS',
  { id: string; name: string }
> = {
  FETCH_ACCOUNT: { id: 'fetch-account', name: 'Fetch Account' },
  FETCH_MACHINES: { id: 'fetch-machines', name: 'Fetch Machines' },
  FETCH_LOGON_USERS: { id: 'fetch-logon-users', name: 'Fetch Logon Users' },
  FETCH_USERS: { id: 'fetch-users', name: 'Fetch Users' },
  FETCH_FINDINGS: {
    id: 'fetch-findings',
    name: 'Fetch Findings',
  },
  FINDING_CVE_RELATIONSHIP: {
    id: 'build-finding-cve-relationship',
    name: 'Build Finding and CVE Relationship',
  },
  FETCH_ENDPOINTS: {
    id: 'fetch-endpoints',
    name: 'Fetch Endpoints',
  },
};

export const Entities: Record<
  'ACCOUNT' | 'MACHINE' | 'LOGON_USER' | 'USER' | 'FINDING' | 'ENDPOINT',
  StepEntityMetadata
> = {
  ACCOUNT: {
    resourceName: 'Account',
    _type: 'microsoft_defender_account',
    _class: ['Account'],
  },
  MACHINE: {
    resourceName: 'Machine',
    _type: 'microsoft_defender_machine',
    _class: ['HostAgent'],
  },
  LOGON_USER: {
    resourceName: 'Logon User',
    _type: 'microsoft_defender_logon_user',
    _class: ['User'],
  },
  USER: {
    resourceName: 'User',
    _type: 'microsoft_defender_user',
    _class: ['User'],
  },
  FINDING: {
    resourceName: 'Finding',
    _type: 'microsoft_defender_finding',
    _class: ['Finding'],
  },
  ENDPOINT: {
    resourceName: 'Device/Machine/Host',
    _type: 'user_endpoint',
    _class: ['Device'],
  },
};

export const Relationships: Record<
  | 'ACCOUNT_HAS_MACHINE'
  | 'ACCOUNT_HAS_USER'
  | 'MACHINE_HAS_LOGON_USER'
  | 'MACHINE_IDENTIFIED_FINDING'
  | 'MACHINE_MANAGES_ENDPOINT',
  StepRelationshipMetadata
> = {
  ACCOUNT_HAS_MACHINE: {
    _type: 'microsoft_defender_account_has_machine',
    sourceType: Entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.MACHINE._type,
  },
  ACCOUNT_HAS_USER: {
    _type: 'microsoft_defender_account_has_user',
    sourceType: Entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.USER._type,
  },
  MACHINE_HAS_LOGON_USER: {
    _type: 'microsoft_defender_machine_has_logon_user',
    sourceType: Entities.MACHINE._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.LOGON_USER._type,
  },
  MACHINE_IDENTIFIED_FINDING: {
    _type: 'microsoft_defender_machine_identified_finding',
    sourceType: Entities.MACHINE._type,
    _class: RelationshipClass.IDENTIFIED,
    targetType: Entities.FINDING._type,
  },
  MACHINE_MANAGES_ENDPOINT: {
    _type: 'microsoft_defender_machine_manages_user_endpoint',
    sourceType: Entities.MACHINE._type,
    _class: RelationshipClass.MANAGES,
    targetType: Entities.ENDPOINT._type,
  },
};

export const MappedRelationships: Record<
  'FINDING_IS_CVE',
  StepMappedRelationshipMetadata
> = {
  FINDING_IS_CVE: {
    _type: 'microsoft_defender_finding_is_cve',
    sourceType: Entities.FINDING._type,
    _class: RelationshipClass.IS,
    targetType: 'cve',
    direction: RelationshipDirection.FORWARD,
  },
};

export const ACCOUNT_ENTITY_KEY = 'entity:account';
