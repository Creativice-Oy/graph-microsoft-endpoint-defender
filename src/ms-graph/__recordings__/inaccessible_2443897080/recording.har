{
  "log": {
    "_recordingName": "inaccessible",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "6.0.5"
    },
    "entries": [
      {
        "_id": "aa33ceaf98a63b8bd52b3986a9ee06cb",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "name": "user-agent",
              "value": "azsdk-js-identity/2.1.0 core-rest-pipeline/1.9.2 Node/v14.17.6 OS/(x64-Darwin-21.6.0)"
            },
            {
              "name": "x-ms-client-request-id",
              "value": "ff27b3eb-9af7-4527-92c6-ca67e5381c9b"
            },
            {
              "name": "host",
              "value": "login.microsoftonline.com"
            }
          ],
          "headersSize": 447,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "api-version",
              "value": "1.1"
            },
            {
              "name": "authorization_endpoint",
              "value": "https://login.microsoftonline.com/ca1b195a-d414-4b2f-8193-34959da8ad09/oauth2/v2.0/authorize"
            }
          ],
          "url": "https://login.microsoftonline.com/common/discovery/instance?api-version=1.1&authorization_endpoint=https%3A%2F%2Flogin.microsoftonline.com%2Fca1b195a-d414-4b2f-8193-34959da8ad09%2Foauth2%2Fv2.0%2Fauthorize"
        },
        "response": {
          "bodySize": 980,
          "content": {
            "mimeType": "application/json; charset=utf-8",
            "size": 980,
            "text": "{\"tenant_discovery_endpoint\":\"https://login.microsoftonline.com/ca1b195a-d414-4b2f-8193-34959da8ad09/v2.0/.well-known/openid-configuration\",\"api-version\":\"1.1\",\"metadata\":[{\"preferred_network\":\"login.microsoftonline.com\",\"preferred_cache\":\"login.windows.net\",\"aliases\":[\"login.microsoftonline.com\",\"login.windows.net\",\"login.microsoft.com\",\"sts.windows.net\"]},{\"preferred_network\":\"login.partner.microsoftonline.cn\",\"preferred_cache\":\"login.partner.microsoftonline.cn\",\"aliases\":[\"login.partner.microsoftonline.cn\",\"login.chinacloudapi.cn\"]},{\"preferred_network\":\"login.microsoftonline.de\",\"preferred_cache\":\"login.microsoftonline.de\",\"aliases\":[\"login.microsoftonline.de\"]},{\"preferred_network\":\"login.microsoftonline.us\",\"preferred_cache\":\"login.microsoftonline.us\",\"aliases\":[\"login.microsoftonline.us\",\"login.usgovcloudapi.net\"]},{\"preferred_network\":\"login-us.microsoftonline.com\",\"preferred_cache\":\"login-us.microsoftonline.com\",\"aliases\":[\"login-us.microsoftonline.com\"]}]}"
          },
          "cookies": [
            {
              "expires": "2022-10-14T05:56:29.000Z",
              "httpOnly": true,
              "name": "fpc",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".login.microsoftonline.com",
              "httpOnly": true,
              "name": "esctx",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "x-ms-gateway-slice",
              "path": "/",
              "sameSite": "none",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "stsservicecookie",
              "path": "/",
              "sameSite": "none",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "cache-control",
              "value": "max-age=86400, private"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "access-control-allow-methods",
              "value": "GET, OPTIONS"
            },
            {
              "name": "p3p",
              "value": "CP=\"DSP CUR OTPi IND OTRi ONL FIN\""
            },
            {
              "name": "x-ms-request-id",
              "value": "ccbb13f3-7421-41dc-99f9-92b844f8c501"
            },
            {
              "name": "x-ms-ests-server",
              "value": "2.1.13562.12 - SEASLR2 ProdSlices"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "date",
              "value": "Wed, 14 Sep 2022 05:56:28 GMT"
            },
            {
              "name": "content-length",
              "value": "980"
            }
          ],
          "headersSize": 1070,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-09-14T05:56:29.289Z",
        "time": 143,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 143
        }
      },
      {
        "_id": "aa33ceaf98a63b8bd52b3986a9ee06cb",
        "_order": 1,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "name": "user-agent",
              "value": "azsdk-js-identity/2.1.0 core-rest-pipeline/1.9.2 Node/v14.17.6 OS/(x64-Darwin-21.6.0)"
            },
            {
              "name": "x-ms-client-request-id",
              "value": "866e18e4-b8ed-470b-99b1-4374cfa999d8"
            },
            {
              "name": "host",
              "value": "login.microsoftonline.com"
            }
          ],
          "headersSize": 350,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://login.microsoftonline.com/ca1b195a-d414-4b2f-8193-34959da8ad09/v2.0/.well-known/openid-configuration"
        },
        "response": {
          "bodySize": 1753,
          "content": {
            "mimeType": "application/json; charset=utf-8",
            "size": 1753,
            "text": "{\"token_endpoint\":\"https://login.microsoftonline.com/ca1b195a-d414-4b2f-8193-34959da8ad09/oauth2/v2.0/token\",\"token_endpoint_auth_methods_supported\":[\"client_secret_post\",\"private_key_jwt\",\"client_secret_basic\"],\"jwks_uri\":\"https://login.microsoftonline.com/ca1b195a-d414-4b2f-8193-34959da8ad09/discovery/v2.0/keys\",\"response_modes_supported\":[\"query\",\"fragment\",\"form_post\"],\"subject_types_supported\":[\"pairwise\"],\"id_token_signing_alg_values_supported\":[\"RS256\"],\"response_types_supported\":[\"code\",\"id_token\",\"code id_token\",\"id_token token\"],\"scopes_supported\":[\"openid\",\"profile\",\"email\",\"offline_access\"],\"issuer\":\"https://login.microsoftonline.com/ca1b195a-d414-4b2f-8193-34959da8ad09/v2.0\",\"request_uri_parameter_supported\":false,\"userinfo_endpoint\":\"https://graph.microsoft.com/oidc/userinfo\",\"authorization_endpoint\":\"https://login.microsoftonline.com/ca1b195a-d414-4b2f-8193-34959da8ad09/oauth2/v2.0/authorize\",\"device_authorization_endpoint\":\"https://login.microsoftonline.com/ca1b195a-d414-4b2f-8193-34959da8ad09/oauth2/v2.0/devicecode\",\"http_logout_supported\":true,\"frontchannel_logout_supported\":true,\"end_session_endpoint\":\"https://login.microsoftonline.com/ca1b195a-d414-4b2f-8193-34959da8ad09/oauth2/v2.0/logout\",\"claims_supported\":[\"sub\",\"iss\",\"cloud_instance_name\",\"cloud_instance_host_name\",\"cloud_graph_host_name\",\"msgraph_host\",\"aud\",\"exp\",\"iat\",\"auth_time\",\"acr\",\"nonce\",\"preferred_username\",\"name\",\"tid\",\"ver\",\"at_hash\",\"c_hash\",\"email\"],\"kerberos_endpoint\":\"https://login.microsoftonline.com/ca1b195a-d414-4b2f-8193-34959da8ad09/kerberos\",\"tenant_region_scope\":\"NA\",\"cloud_instance_name\":\"microsoftonline.com\",\"cloud_graph_host_name\":\"graph.windows.net\",\"msgraph_host\":\"graph.microsoft.com\",\"rbac_url\":\"https://pas.windows.net\"}"
          },
          "cookies": [
            {
              "expires": "2022-10-14T05:56:29.000Z",
              "httpOnly": true,
              "name": "fpc",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "domain": ".login.microsoftonline.com",
              "httpOnly": true,
              "name": "esctx",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "x-ms-gateway-slice",
              "path": "/",
              "sameSite": "none",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "stsservicecookie",
              "path": "/",
              "sameSite": "none",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "cache-control",
              "value": "max-age=86400, private"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "access-control-allow-methods",
              "value": "GET, OPTIONS"
            },
            {
              "name": "p3p",
              "value": "CP=\"DSP CUR OTPi IND OTRi ONL FIN\""
            },
            {
              "name": "x-ms-request-id",
              "value": "d07bd38b-031a-449d-b8aa-aeaeb9701800"
            },
            {
              "name": "x-ms-ests-server",
              "value": "2.1.13672.7 - WUS2 ProdSlices"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "date",
              "value": "Wed, 14 Sep 2022 05:56:29 GMT"
            },
            {
              "name": "content-length",
              "value": "1753"
            }
          ],
          "headersSize": 1067,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-09-14T05:56:29.437Z",
        "time": 340,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 340
        }
      },
      {
        "_id": "b4391da807354cda8b8ffaab25e22621",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 559,
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/x-www-form-urlencoded;charset=utf-8"
            },
            {
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "name": "user-agent",
              "value": "azsdk-js-identity/2.1.0 core-rest-pipeline/1.9.2 Node/v14.17.6 OS/(x64-Darwin-21.6.0)"
            },
            {
              "name": "x-ms-client-request-id",
              "value": "b42148ad-ad6a-4dab-bb16-5d59e4ffb4dc"
            },
            {
              "name": "content-length",
              "value": "559"
            },
            {
              "name": "host",
              "value": "login.microsoftonline.com"
            }
          ],
          "headersSize": 415,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "postData": {
            "mimeType": "application/x-www-form-urlencoded;charset=utf-8",
            "params": [],
            "text": "[REDACTED]"
          },
          "queryString": [],
          "url": "https://login.microsoftonline.com/ca1b195a-d414-4b2f-8193-34959da8ad09/oauth2/v2.0/token"
        },
        "response": {
          "bodySize": 632,
          "content": {
            "mimeType": "application/json; charset=utf-8",
            "size": 632,
            "text": "{\"error\":\"invalid_client\",\"error_description\":\"AADSTS7000215: Invalid client secret provided. Ensure the secret being sent in the request is the client secret value, not the client secret ID, for a secret added to app '3d7e133b-9537-49e5-9006-1408cda46f55'.\\r\\nTrace ID: 0e33fb29-e5d4-4894-9a2a-e5f0c3874200\\r\\nCorrelation ID: b42148ad-ad6a-4dab-bb16-5d59e4ffb4dc\\r\\nTimestamp: 2022-09-14 05:56:30Z\",\"error_codes\":[7000215],\"timestamp\":\"2022-09-14 05:56:30Z\",\"trace_id\":\"0e33fb29-e5d4-4894-9a2a-e5f0c3874200\",\"correlation_id\":\"b42148ad-ad6a-4dab-bb16-5d59e4ffb4dc\",\"error_uri\":\"https://login.microsoftonline.com/error?code=7000215\"}"
          },
          "cookies": [
            {
              "expires": "2022-10-14T05:56:30.000Z",
              "httpOnly": true,
              "name": "fpc",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "x-ms-gateway-slice",
              "path": "/",
              "sameSite": "none",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "stsservicecookie",
              "path": "/",
              "sameSite": "none",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "cache-control",
              "value": "no-store, no-cache"
            },
            {
              "name": "pragma",
              "value": "no-cache"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "name": "expires",
              "value": "-1"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "p3p",
              "value": "CP=\"DSP CUR OTPi IND OTRi ONL FIN\""
            },
            {
              "name": "x-ms-request-id",
              "value": "0e33fb29-e5d4-4894-9a2a-e5f0c3874200"
            },
            {
              "name": "x-ms-ests-server",
              "value": "2.1.13622.7 - EUS ProdSlices"
            },
            {
              "name": "x-ms-clitelem",
              "value": "1,7000215,0,,"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "date",
              "value": "Wed, 14 Sep 2022 05:56:29 GMT"
            },
            {
              "name": "content-length",
              "value": "632"
            }
          ],
          "headersSize": 765,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 401,
          "statusText": "Unauthorized"
        },
        "startedDateTime": "2022-09-14T05:56:29.787Z",
        "time": 386,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 386
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
