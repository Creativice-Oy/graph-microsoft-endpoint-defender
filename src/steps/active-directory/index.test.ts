import { executeStepWithDependencies } from '@jupiterone/integration-sdk-testing';
import { Recording, setupProjectRecording } from '../../../test/recording';
import { buildStepTestConfigForStep } from '../../../test/config';
import { Steps } from '../../constants';

// See test/README.md for details
let recording: Recording;
afterEach(async () => {
  await recording.stop();
});

test('#fetchAccount', async () => {
  recording = setupProjectRecording({
    directory: __dirname,
    name: '#fetchAccount',
  });

  const stepConfig = buildStepTestConfigForStep(Steps.FETCH_ACCOUNT.id);
  const stepResult = await executeStepWithDependencies(stepConfig);
  expect(stepResult).toMatchStepMetadata(stepConfig);
  expect(stepResult).toMatchSnapshot();
});

test('#fetchUsers', async () => {
  recording = setupProjectRecording({
    directory: __dirname,
    name: '#fetchUsers',
  });

  const stepConfig = buildStepTestConfigForStep(Steps.FETCH_USERS.id);
  const stepResult = await executeStepWithDependencies(stepConfig);
  expect(stepResult).toMatchStepMetadata(stepConfig);
  expect(stepResult).toMatchSnapshot();
});
