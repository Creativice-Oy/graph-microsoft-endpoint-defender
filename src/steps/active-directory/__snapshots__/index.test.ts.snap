// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchAccount 1`] = `
Object {
  "collectedData": Object {
    "entity:account": Object {
      "_class": Array [
        "Account",
      ],
      "_key": "microsoft_defender_account:local-integration-instance",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "organization": Object {
              "assignedPlans": Array [
                Object {
                  "assignedDateTime": "2022-09-13T09:30:34Z",
                  "capabilityStatus": "Enabled",
                  "service": "WindowsDefenderATP",
                  "servicePlanId": "36810a13-b903-490a-aa45-afbeb7540832",
                },
                Object {
                  "assignedDateTime": "2022-09-12T12:14:50Z",
                  "capabilityStatus": "Enabled",
                  "service": "exchange",
                  "servicePlanId": "113feb6c-3fe4-4440-bddc-54d774bf0318",
                },
                Object {
                  "assignedDateTime": "2022-09-12T12:14:50Z",
                  "capabilityStatus": "Enabled",
                  "service": "WindowsDefenderATP",
                  "servicePlanId": "871d91ec-ec1a-452b-a83f-bd76c7d770ef",
                },
                Object {
                  "assignedDateTime": "2022-09-12T12:14:50Z",
                  "capabilityStatus": "Enabled",
                  "service": "SCO",
                  "servicePlanId": "1689aade-3d6a-4bfc-b017-46d2672df5ad",
                },
              ],
              "businessPhones": Array [
                "5859430080",
              ],
              "city": null,
              "country": null,
              "countryLetterCode": "US",
              "createdDateTime": "2022-09-12T12:14:35Z",
              "deletedDateTime": null,
              "directorySizeQuota": Object {
                "total": 50000,
                "used": 112,
              },
              "displayName": "JupiterOne",
              "id": "ca1b195a-d414-4b2f-8193-34959da8ad09",
              "isMultipleDataLocationsForServicesEnabled": null,
              "marketingNotificationEmails": Array [],
              "onPremisesLastSyncDateTime": null,
              "onPremisesSyncEnabled": null,
              "postalCode": null,
              "preferredLanguage": "en",
              "privacyProfile": null,
              "provisionedPlans": Array [
                Object {
                  "capabilityStatus": "Enabled",
                  "provisioningStatus": "Success",
                  "service": "SCO",
                },
                Object {
                  "capabilityStatus": "Enabled",
                  "provisioningStatus": "Success",
                  "service": "exchange",
                },
                Object {
                  "capabilityStatus": "Enabled",
                  "provisioningStatus": "Success",
                  "service": "exchange",
                },
              ],
              "securityComplianceNotificationMails": Array [],
              "securityComplianceNotificationPhones": Array [],
              "state": null,
              "street": null,
              "technicalNotificationMails": Array [
                "austin.kelleher@jupiterone.dev",
              ],
              "tenantType": "AAD",
              "verifiedDomains": Array [
                Object {
                  "capabilities": "Email, OfficeCommunicationsOnline",
                  "isDefault": true,
                  "isInitial": true,
                  "name": "jupiteroneintdev.onmicrosoft.com",
                  "type": "Managed",
                },
              ],
            },
          },
        },
      ],
      "_type": "microsoft_defender_account",
      "createdOn": undefined,
      "defaultDomain": "jupiteroneintdev.onmicrosoft.com",
      "displayName": "Local Integration",
      "id": "ca1b195a-d414-4b2f-8193-34959da8ad09",
      "name": "JupiterOne",
      "organizationName": "JupiterOne",
      "verifiedDomains": Array [
        "jupiteroneintdev.onmicrosoft.com",
      ],
    },
  },
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "microsoft_defender_account:local-integration-instance",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "organization": Object {
              "assignedPlans": Array [
                Object {
                  "assignedDateTime": "2022-09-13T09:30:34Z",
                  "capabilityStatus": "Enabled",
                  "service": "WindowsDefenderATP",
                  "servicePlanId": "36810a13-b903-490a-aa45-afbeb7540832",
                },
                Object {
                  "assignedDateTime": "2022-09-12T12:14:50Z",
                  "capabilityStatus": "Enabled",
                  "service": "exchange",
                  "servicePlanId": "113feb6c-3fe4-4440-bddc-54d774bf0318",
                },
                Object {
                  "assignedDateTime": "2022-09-12T12:14:50Z",
                  "capabilityStatus": "Enabled",
                  "service": "WindowsDefenderATP",
                  "servicePlanId": "871d91ec-ec1a-452b-a83f-bd76c7d770ef",
                },
                Object {
                  "assignedDateTime": "2022-09-12T12:14:50Z",
                  "capabilityStatus": "Enabled",
                  "service": "SCO",
                  "servicePlanId": "1689aade-3d6a-4bfc-b017-46d2672df5ad",
                },
              ],
              "businessPhones": Array [
                "5859430080",
              ],
              "city": null,
              "country": null,
              "countryLetterCode": "US",
              "createdDateTime": "2022-09-12T12:14:35Z",
              "deletedDateTime": null,
              "directorySizeQuota": Object {
                "total": 50000,
                "used": 112,
              },
              "displayName": "JupiterOne",
              "id": "ca1b195a-d414-4b2f-8193-34959da8ad09",
              "isMultipleDataLocationsForServicesEnabled": null,
              "marketingNotificationEmails": Array [],
              "onPremisesLastSyncDateTime": null,
              "onPremisesSyncEnabled": null,
              "postalCode": null,
              "preferredLanguage": "en",
              "privacyProfile": null,
              "provisionedPlans": Array [
                Object {
                  "capabilityStatus": "Enabled",
                  "provisioningStatus": "Success",
                  "service": "SCO",
                },
                Object {
                  "capabilityStatus": "Enabled",
                  "provisioningStatus": "Success",
                  "service": "exchange",
                },
                Object {
                  "capabilityStatus": "Enabled",
                  "provisioningStatus": "Success",
                  "service": "exchange",
                },
              ],
              "securityComplianceNotificationMails": Array [],
              "securityComplianceNotificationPhones": Array [],
              "state": null,
              "street": null,
              "technicalNotificationMails": Array [
                "austin.kelleher@jupiterone.dev",
              ],
              "tenantType": "AAD",
              "verifiedDomains": Array [
                Object {
                  "capabilities": "Email, OfficeCommunicationsOnline",
                  "isDefault": true,
                  "isInitial": true,
                  "name": "jupiteroneintdev.onmicrosoft.com",
                  "type": "Managed",
                },
              ],
            },
          },
        },
      ],
      "_type": "microsoft_defender_account",
      "createdOn": undefined,
      "defaultDomain": "jupiteroneintdev.onmicrosoft.com",
      "displayName": "Local Integration",
      "id": "ca1b195a-d414-4b2f-8193-34959da8ad09",
      "name": "JupiterOne",
      "organizationName": "JupiterOne",
      "verifiedDomains": Array [
        "jupiteroneintdev.onmicrosoft.com",
      ],
    },
  ],
  "collectedRelationships": Array [],
  "encounteredEntityKeys": Set {
    "microsoft_defender_account:local-integration-instance",
  },
  "encounteredTypes": Array [
    "microsoft_defender_account",
  ],
}
`;

exports[`#fetchUsers 1`] = `
Object {
  "collectedData": Object {},
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "microsoft_defender_user:53c5938b-5d0f-4861-9f54-268fc6f28b1b",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "businessPhones": Array [
              "5859430080",
            ],
            "displayName": "Austin Kelleher",
            "givenName": "Austin",
            "id": "53c5938b-5d0f-4861-9f54-268fc6f28b1b",
            "jobTitle": null,
            "mail": null,
            "mobilePhone": null,
            "officeLocation": null,
            "preferredLanguage": null,
            "surname": "Kelleher",
            "userPrincipalName": "austin.kelleher@jupiteroneintdev.onmicrosoft.com",
          },
        },
      ],
      "_type": "microsoft_defender_user",
      "active": true,
      "businessPhones": Array [
        "5859430080",
      ],
      "createdOn": undefined,
      "displayName": "Austin Kelleher",
      "givenName": "Austin",
      "id": "53c5938b-5d0f-4861-9f54-268fc6f28b1b",
      "jobTitle": null,
      "mail": null,
      "mobilePhone": null,
      "name": "Austin Kelleher",
      "officeLocation": null,
      "preferredLanguage": null,
      "surname": "Kelleher",
      "userPrincipalName": "austin.kelleher@jupiteroneintdev.onmicrosoft.com",
      "username": "austin.kelleher@jupiteroneintdev.onmicrosoft.com",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "microsoft_defender_user:e92080d6-6611-48dc-8241-3046f8f20a43",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "businessPhones": Array [],
            "displayName": "Stefan Virag",
            "givenName": "Stefan",
            "id": "e92080d6-6611-48dc-8241-3046f8f20a43",
            "jobTitle": null,
            "mail": null,
            "mobilePhone": null,
            "officeLocation": null,
            "preferredLanguage": null,
            "surname": "Virag",
            "userPrincipalName": "stefan.virag@jupiteroneintdev.onmicrosoft.com",
          },
        },
      ],
      "_type": "microsoft_defender_user",
      "active": true,
      "businessPhones": Array [],
      "createdOn": undefined,
      "displayName": "Stefan Virag",
      "givenName": "Stefan",
      "id": "e92080d6-6611-48dc-8241-3046f8f20a43",
      "jobTitle": null,
      "mail": null,
      "mobilePhone": null,
      "name": "Stefan Virag",
      "officeLocation": null,
      "preferredLanguage": null,
      "surname": "Virag",
      "userPrincipalName": "stefan.virag@jupiteroneintdev.onmicrosoft.com",
      "username": "stefan.virag@jupiteroneintdev.onmicrosoft.com",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "microsoft_defender_account:local-integration-instance",
      "_key": "microsoft_defender_account:local-integration-instance|has|microsoft_defender_user:53c5938b-5d0f-4861-9f54-268fc6f28b1b",
      "_toEntityKey": "microsoft_defender_user:53c5938b-5d0f-4861-9f54-268fc6f28b1b",
      "_type": "microsoft_defender_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "microsoft_defender_account:local-integration-instance",
      "_key": "microsoft_defender_account:local-integration-instance|has|microsoft_defender_user:e92080d6-6611-48dc-8241-3046f8f20a43",
      "_toEntityKey": "microsoft_defender_user:e92080d6-6611-48dc-8241-3046f8f20a43",
      "_type": "microsoft_defender_account_has_user",
      "displayName": "HAS",
    },
  ],
  "encounteredEntityKeys": Set {
    "microsoft_defender_user:53c5938b-5d0f-4861-9f54-268fc6f28b1b",
    "microsoft_defender_user:e92080d6-6611-48dc-8241-3046f8f20a43",
    "microsoft_defender_account:local-integration-instance",
  },
  "encounteredTypes": Array [
    "microsoft_defender_account",
    "microsoft_defender_user",
    "microsoft_defender_account_has_user",
  ],
}
`;
